File: CodeForces/Competitions/1333 - Round #632 Div 2/1333C_TLE_eugene_and_an_array.cpp
// https://codeforces.com/problemset/problem/1333/C
// 1333C Eugene and an Array


#include <iostream>
#include <map>

#define mp make_pair

using namespace std;

int total = 0;

map<pair<int, int>, bool> dp;
long long arr[200001];

bool checkSubarray(int begin, int end) {
    auto itr = dp.find(mp(begin, end));
    if (itr != dp.end())
        return itr->second;
    else if (end-begin == 1) {
        bool isNotZero = arr[begin] != (begin > 0 ? arr[begin-1] : 0);
        dp.insert({mp(begin, end), isNotZero});
        return isNotZero;
    }
    else {
        bool res = checkSubarray(begin, end-1) && checkSubarray(begin+1, end);
        bool nres = arr[end-1] != (begin-1 < 0 ? 0 : arr[begin-1]);
        res &= nres;
        dp.insert({mp(begin,end), res});
        total += res ? 1 : 0;
        return res;
    }
}


int main() {
    
    int L;
    cin >> L;

    cin >> arr[0];
    unsigned long count = 0;
    if (arr[0] == 0) count++;
    for (int i=1; i<L; ++i) {
        cin >> arr[i];
        if (arr[i] == 0) count++;
        arr[i] += arr[i-1];
    }

    total = L - count;

    if (total == 0) {
        cout << 0;
        return 0;
    }

    auto res = checkSubarray(0, L);
    
    cout << total;
}

File: CodeForces/Competitions/1333 - Round #632 Div 2/1333C_TLE_eugene_and_an_array.py
# https://codeforces.com/problemset/problem/1333/C
# 1333C Eugene and an Array

''' TOO SLOW (TLE) '''

L = int(input())
arr = [int(x) for x in input().split(" ")]

initial = 0
while(initial < len(arr) and arr[initial] == 0):
    initial += 1
final = len(arr)-1
while(final >= 0 and arr[final] == 0):
    final -= 1
if final < initial:
    print(0)
    exit()

arr = arr[initial:final+1]

L = len(arr)
total = L - arr.count(0)

dp = []
for x in range(L+1):
    dp.append([])
    for i in range(L+1):
        dp[x].append(None)


for i in range(len(arr)):
    arr[i] = arr[i] + (arr[i-1] if i-1 >= 0 else 0)


def checkSubarray(begin, end):
    global total
    if dp[begin][end] is not None:
        return dp[begin][end]
    elif (len(arr[begin:end]) == 1):
        isNotZero = arr[begin] != (arr[begin-1] if begin-1 >= 0 else 0)
        dp[begin][end] = isNotZero
        return isNotZero
    else:
        res = checkSubarray(begin, end-1) and checkSubarray(begin+1, end)
        new = arr[end-1] != (0 if begin-1 < 0 else arr[begin-1])
        res &= new
        dp[begin][end] = res
        total += 1 if res else 0
        return res


checkSubarray(0, len(arr))

print(total)
